File: Recursos/FISH 559/Day 2/Ex2.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(Nprey)
  DATA_INTEGER(NData)
  DATA_MATRIX(N);
  DATA_INTEGER(Model_Num);

  PARAMETER_VECTOR(alpha);
  PARAMETER_VECTOR(beta);
  PARAMETER_VECTOR(gamma);

  matrix<Type> Pred(NData,Nprey);
  Type obj_fun;

  int Idata,Iprey;

  for (Idata=0;Idata<NData;Idata++)
   for (Iprey=0;Iprey<Nprey;Iprey++)
    {
     // Model 1
     if (Model_Num == 1) Pred(Idata,Iprey) = alpha(Iprey)*N(Idata,0);

     // Model 2
     if (Model_Num == 2) Pred(Idata,Iprey) = alpha(Iprey)*N(Idata,0)/(1.0+beta(Iprey)*N(Idata,1+Iprey));

     // Model 3
     if (Model_Num == 3) Pred(Idata,Iprey) = alpha(Iprey)*N(Idata,0)*pow(N(Idata,1+Iprey),gamma(Iprey)-1.0)/(1.0+beta(Iprey)*pow(N(Idata,1+Iprey),gamma(Iprey)));

     // Model 4
     if (Model_Num == 4) Pred(Idata,Iprey) = alpha(Iprey)*N(Idata,0)/(1+beta(Iprey)*N(Idata,1+Iprey)+gamma(Iprey)*N(Idata,0));

     // Residuals
     obj_fun = obj_fun + pow(log(Pred(Idata,Iprey))-log(N(Idata,4+Iprey)),2.0);
    }
  return(obj_fun);

}

File: Recursos/FISH 559/Day 2/ex2.R
setwd("D:\\courses\\FISH 559_20\\TMB Workshop\\In Class Assignments\\Ex2")

require(TMB)
compile("Ex2.cpp", flags="-Wno-ignored-attributes")
dyn.load(dynlib("Ex2"))

################################################################################

Nprey <- scan("ex2.dat",skip=1,n=1,quiet=T)
Ndata <- scan("ex2.dat",skip=3,n=1,quiet=T)
dataD <- matrix(scan("ex2.dat",skip=5,n=7*Ndata,quiet=T),ncol=7,byrow=T)
Nprey <- scan("ex2.dat",skip=1,n=1,quiet=T)

Model_Num <- 4
data <- list(Nprey=Nprey,NData=Ndata,N=dataD,Model_Num=Model_Num)
parameters <- list(alpha=rep(0.1,3),beta=rep(0,3),gamma=rep(0,3))

# map to help decide which parameters to estimates
if (Model_Num==1) map <- list(beta=factor(rep(NA,Nprey)),gamma=factor(rep(NA,Nprey)))
if (Model_Num==2) map <- list(gamma=factor(rep(NA,Nprey)))
if (Model_Num==3) map <- NULL
if (Model_Num==4) map <- NULL

#print(data)
#print(parameters)
#AAA

################################################################################

model <- MakeADFun(data, parameters, DLL="Ex2",silent=T,map=map)
fit <- nlminb(model$par, model$fn, model$gr)
best <- model$env$last.par.best
rep <- sdreport(model)
print(best)
print(rep)
